Sources/AsyncAlgorithms/MultiProducerSingleConsumerChannel/MultiProducerSingleConsumerChannel+Internal.swift

```diff
@@ -0,0 +1,1409 @@
+                for producerContinuation in producerContinuations {
+                    switch producerContinuation {
+                    case .closure(let onProduceMore):
+                        onProduceMore(.failure(MultiProducerSingleConsumerChannelAlreadyFinishedError()))
+                    case .continuation(let continuation):
+                        continuation.resume(throwing: MultiProducerSingleConsumerChannelAlreadyFinishedError())
+                    }
+                }
+
+            case .none:
+                break
+            }
+        }
+
+        @inlinable
+        func next(isolation actor: isolated (any Actor)?) async throws -> Element? {
+            let action = self._lock.withLock {
+                self._stateMachine.next()
+            }
+
+            switch action {
+            case .returnElement(let element):
+                return element
+
+            case .returnElementAndResumeProducers(let element, let producerContinuations):
+                for producerContinuation in producerContinuations {
+                    switch producerContinuation {
+                    case .closure(let onProduceMore):
+                        onProduceMore(.success(()))
+                    case .continuation(let continuation):
+                        continuation.resume()
+                    }
+                }
+
+                return element
+
+            case .returnFailureAndCallOnTermination(let failure, let onTermination):
+                onTermination?()
+                switch failure {
+                case .some(let error):
+                    throw error
+
+                case .none:
+                    return nil
+                }
+
+            case .returnNil:
+                return nil
+
+            case .suspendTask:
+                return try await self.suspendNext(isolation: actor)
+            }
+        }
+
+        @inlinable
+        func suspendNext(isolation actor: isolated (any Actor)?) async throws -> Element? {
+            return try await withTaskCancellationHandler {
+                return try await withUnsafeThrowingContinuation { continuation in
+                    let action = self._lock.withLock {
+                        self._stateMachine.suspendNext(continuation: continuation)
+                    }
+
+                    switch action {
+                    case .resumeConsumerWithElement(let continuation, let element):
+                        continuation.resume(returning: element)
+
+                    case .resumeConsumerWithElementAndProducers(let continuation, let element, let producerContinuations):
+                        continuation.resume(returning: element)
+                        for producerContinuation in producerContinuations {
+                            switch producerContinuation {
+                            case .closure(let onProduceMore):
+                                onProduceMore(.failure(CancellationError()))
```

Comment:
This is a state machine bug I picked up running the OpenAPIURLSession tests with this type:

```suggestion
                                onProduceMore(.success(()))
```

Might be worth seeing if we can have similar tests in this project. This particular test, `1kChunk_10kChunks_100BDownloadWatermark`, exercised the suspend and resume of the producer a lot due to the relation of the chunk size to the watermark.